doc/designs/fips_indicator.md

```diff
@@ -23,215 +26,62 @@ The following information was extracted from the FIPS 140-3 IG [5] “2.4.C Appr
-- Cons: The locking increment seems messy.
-
-### Proposed Solution (Using an indicator everywhere)
-
-Add a OSSL_PARAM getter to each provider algorithm context.
-By default if the getter is not handled then it would return not approved.
-
-- Pros: The code is easier to find since it is part of the get_ctx_params function.
-- Pros: The getter can be called at any point after the setting is done.
-
-Any fips algorithm that is approved would then need a setter that at a minimum contains code similar to the following
-
-``` C
-int ossl_xxx_fips_approved(void)
-{
-ifdef FIPS_MODULE
-    return 1; // conditional code would go here for each algorithm if required *
-else
-    return 0;
-endif
-}
-```
-
-and in the algorithms get_ctx() function
-
-``` C
-int xxx_get_fips_approved(OSSL_PARAM params[])
-{
-    p = OSSL_PARAM_locate(params, OSSL_FIPS_PARAM_APPROVED);
-    if (p != NULL && !OSSL_PARAM_set_int(p, ossl_xxx_fips_approved()))
-        return 0;
-    return 1;
-}
-```
-
-### Questions ###
-
-- Do we want to use 3 states? (Approved, Not Approved, Undetermined)
-- Do we want this to apply this only to the FIPS provider?
-
-### API’s that would be used to support this are
-
-- EVP_PKEY Keygen, Encryption, Signatures, Key Exchange, KEM
-
-``` C
-EVP_PKEY_CTX_get_params(ctx, );
-```
-
-(Note that this would mean you could not use functions that hide the ctx such as EVP_PKEY_Q_keygen()!)
-
-- Ciphers
-
-``` C
-EVP_CIPHER_CTX_get_params()
-```
-
-- Digests
-
-``` C
-EVP_MD_CTX_get_params()
-```
-
-- KDF’s
-
-``` C
-EVP_KDF_CTX_get_params()
-```
-
-- MAC’s
-
-``` C
-EVP_MAC_CTX_get_params()
-```
-
-- RAND
-
-``` C
-EVP_RAND_CTX_get_params()
-```
+Solution
+--------
+
+We already have most of the existing code in the FIPS provider using
+implicit indicators i.e. An error occurs if existing FIPS rules are violated.
+
+The following rules will apply to any code that currently is not FIPS approved,
+but needs to be.
+
+- There will be a 'fips approved' mode per provider operation context that is on by default.
+- In this mode any FIPS rule that is violated will result in an error.
+- The 'fips approved' mode may be turned off per context via a OSSL_PARAM setter, if it
+is turned off then the operation is not FIPS approved.
+- A getter is required that returns if the operation is 'fips approved' using an
+OSSL_PARAM. The returned value can be a combination of the set 'fips_approved' mode
+plus any other logic. This is an explicit indicator.
+- Any algorithm that is transitioning [4] to not being allowed should be removed from
+the fips provider.
+- If an algorithm is transitioning [4] to be only allowed for processing
+(e.g. verification, signing, key validation) then the protection code
+(keygen, signing, encryption) should be removed from the fips provider, any
+attempt to use the protection API's should result in an error. The processing
+code should still be functional.
```

Comment:
I am not sure we discussed this so strictly in the OTC discussion. Perhaps these algorithms should be kept but work only if you set the non-approved setter? Or maybe this should be a matter of individual decision based on the concrete algorithm that is being removed?

Reply:
Or, we keep the algorithm but just with fips=no property.

In case just the _protection_ functions are being disallowed, IMO we should keep them but operate them only with the non-approved setter set.

Reply:
I am failing to see the point of allowing them.
I really dislike the fips=no option, this implies there would be an entry for fips=no as well as fips=yes, not specifying the query then ends up with either one. What is the motivation for using non FIPS operations such as DSA in a FIPS provider?

Reply:
Historically we've used the `=no` option for disallowed algorithms (for 3DES).  This was in the name of compatibility for existing users of the FIPS provider.  The more I've thought about this, the worse this decision feels.

It makes more sense, to me, for the FIPS provider to only include allowed operations.  This does mean that customers will break when updating but any customer who wants FIPS will also want to stay FIPS.